keps/sig-cli/NNNN-kustomize-components.md

```diff
@@ -0,0 +1,676 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+(**FIXME:** Explain what is the current situation and then roughly explain what this feature does)
+
+> This section is incredibly important for producing high quality user-focused
+> documentation such as release notes or a development roadmap.  It should be
+> possible to collect this information before implementation begins in order to
+> avoid requiring implementors to split their attention between writing release
+> notes and implementing the feature itself.  KEP editors, SIG Docs, and SIG PM
+> should help to ensure that the tone and content of the `Summary` section is
+> useful for a wide audience.
+>
+> A good summary is probably at least a paragraph in length.
+
+## Motivation
+
+By design, Kustomize and its variant system follow an inheritance model rather
+than a composition one. This poses certain limitations in complex deployment
+scenarios that deal with multivariate applications, i.e., ones that:
+
+- ship with numerous opt-in features and corresponding configuration options
+- target dissimilar audiences with different needs and scopes (developers,
+  admins, customers, testers etc.)
+- run on multiple cloud platforms (on-prem, Minikube, GKE, EKS, AKS, etc.)
+- tailor semantically distinct parts of a
+  common, parent configuration simultaneously
+
+Currently, declarative management of such applications with stock Kubernetes
+CLI tools is quite hard and can quickly get out of control.
+
+Zooming in, the mechanism that Kustomize uses to interpret and ultimately apply
+patches does not allow for mutating different aspects of the very same base
+resource in parallel without modifying its GVKN (unique identifier). That is,
+sibling variants that operate on the same level of the hierarchy and are not
+chained (i.e., one does not import the other as base) cannot modify their
+common parent due to resource ID conflicts.
+
+Aside, representation and management of applications with a higher degree of
+freedom with an inheritance model seems rather inefficient. The reason for this
+is that some applications rely on mixins to provide as many off-the-shelf
+combinations (sets of manifests) as possible which, foreseeably, leads to
+duplicate YAML documents and logic in multiple places. This greatly hinders
+reusability of variants across deployments and, in the long term, readability
+and maintainability. Ideally, we would like to reduce repetition of variants
+and treat each variant as a single, unambiguous element according to the
+[DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle.
+
+### Goals
+
+- Introduce the notion of components in ``kustomize`` and, eventually in
+  ``kubectl``
+- Provide the implementation that allows users to define components, i.e.,
+  portable variants that are able to modify a common, base resources without
```

Comment:
a common, base resources -> a set of base resources

Reply:
ACK.